diff -rupN eper/src/prf_crypto.erl eper-new/src/prf_crypto.erl
--- eper/src/prf_crypto.erl	2014-01-15 04:16:27.000000000 +0800
+++ eper-new/src/prf_crypto.erl	2014-01-15 04:20:02.000000000 +0800
@@ -15,17 +15,19 @@ encrypt(Data) -> encrypt(phrase(),Data).
 encrypt(Phrase,Data) ->
   assert_crypto(),
   {Key,Ivec} = make_key(Phrase),
-  crypto:des_cbc_encrypt(Key,Ivec,pad(Data)).
+  %crypto:des_cbc_encrypt(Key,Ivec,pad(Data)).
+  crypto:block_encrypt(des_cbc, Key,Ivec,pad(Data)).
 
 decrypt(Data) -> decrypt(phrase(),Data).
 
 decrypt(Phrase,Data) ->
   assert_crypto(),
   {Key,Ivec} = make_key(Phrase),
-  unpad(crypto:des_cbc_decrypt(Key,Ivec,Data)).
+  %unpad(crypto:des_cbc_decrypt(Key,Ivec,Data)).
+  unpad(crypto:block_decrypt(des_cbc, Key,Ivec,Data)).
 
 make_key(Phrase) ->
-  <<Key:8/binary,Ivec:8/binary>> = crypto:md5(Phrase),
+  <<Key:8/binary,Ivec:8/binary>> = crypto:hash(md5, Phrase),
   {Key,Ivec}.
 
 pad(Term) ->
